### PR TITLE
Support for PowerShell CLM

### PR DESCRIPTION
**Pull request by timbrigham-oc**
Pull request for support of Constrained Language Mode for the ADO pipeline agent. 

This refactors the code out of the CLI and into a script that can be signed, verified and worked with using WDAC and CLM. 

This corresponds to a bug I opened on this issue, #4492.

### CODE REVIEW

**Comment by timbrigham-oc** [comment]
@microsoft-github-policy-service agree company="OneCall"

Requested and received permission for submitting this bug fix from the current CSO.

File: src/dev.sh [hunk 1]
```diff
@@ -361,7 +361,7 @@ bash ./Misc/externals.sh $RUNTIME_ID "Pre-Cache" || checkRC "externals.sh Pre-Ca
 
 if [[ "$CURRENT_PLATFORM" == 'windows' ]]; then
     vswhere=$(find "$DOWNLOAD_DIR" -name vswhere.exe | head -1)
-    vs_location=$("$vswhere" -latest -property installationPath)
+    vs_location=$("$vswhere" -products 'Microsoft.VisualStudio.Product.BuildTools' -latest -property installationPath)
```
Comment (timbrigham-oc) [inline src/dev.sh:364]: Not sure if this should be kept or not, but without a product definition this script doesn't work with VS Code and only the build agents installed from Visual Studio.

Comment (timbrigham-oc) [reply src/dev.sh]: I updated my edited logic here to only try adding the product definition if vs_location is empty, should be no conflict in finding binaries under existing paths

File: src/Agent.Worker/Handlers/PowerShell3Handler.cs [hunk 2]
```diff
@@ -46,8 +46,17 @@ public async Task RunAsync()
...
-                @"-NoLogo -Sta -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command "". ([scriptblock]::Create('if ([Console]::InputEncoding -is [Text.UTF8Encoding] -and [Console]::InputEncoding.GetPreamble().Length -ne 0) {{ [Console]::InputEncoding = New-Object Text.UTF8Encoding $false }} if (!$PSHOME) {{ $null = Get-Item -LiteralPath ''variable:PSHOME'' }} else {{ Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1'')) ; Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1'')) }}')) 2>&1 | ForEach-Object {{ Write-Verbose $_.Exception.Message -Verbose }} ; Import-Module -Name '{0}' -ArgumentList @{{ NonInteractive = $true }} -ErrorAction Stop ; $VerbosePreference = '{1}' ; $DebugPreference = '{1}' ; Invoke-VstsTaskScript -ScriptBlock ([scriptblock]::Create('. ''{2}'''))""",
+                @"-NoLogo -Sta -NoProfile -ExecutionPolicy Unrestricted -Command ""..\..\..\..\bin\powershell\Use-VstsTaskScript.ps1"" -Name {0} -DebugOption {1} -ScriptBlock ""{2}""",
```
Comment (KonstantinTyukalov) [inline src/Agent.Worker/Handlers/PowerShell3Handler.cs:59]: Please don't use relative paths.
I suggest to build path to script using constant and context, e.g. `Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Bin), "powershell", "Use-VstsTaskScript.ps1");`.
And please rename `Use-VstsTaskScript.ps1` => `Start-AzpTask.ps1` (me are moving away from the old VSTS brand)

File: src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1 [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+since it is parameterized it can be signed and trusted for WDAC and CLM.
+This is the original code that has been repalced - 
+@". ([scriptblock]::Create('if ([Console]::InputEncoding -is [Text.UTF8Encoding] -and [Console]::InputEncoding.GetPreamble().Length -ne 0) {{ [Console]::InputEncoding = New-Object Text.UTF8Encoding $false }} if (!$PSHOME) {{ $null = Get-Item -LiteralPath ''variable:PSHOME'' }} else {{ Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1'')) ; Import-Module -Name ([System.IO.Path]::Combine($PSHOME, ''Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1'')) }}')) 2>&1 | ForEach-Object {{ Write-Verbose $_.Exception.Message -Verbose }} ; Import-Module -Name '{0}' -ArgumentList @{{ NonInteractive = $true }} -ErrorAction Stop ; $VerbosePreference = '{1}' ; $DebugPreference = '{1}' ; Invoke-VstsTaskScript -ScriptBlock ([scriptblock]::Create('. ''{2}'''))",
+#>
+
+param ( 
+    [Parameter(mandatory = $true)]$name, # Argument 0 
+    [Parameter(mandatory = $true)]$DebugOption, # Argument 1
+    [Parameter(mandatory = $true)]$scriptBlockString	# Argument 2
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1:12]: The params names don't seem to match the place it is called from 🙂 
https://github.com/microsoft/azure-pipelines-agent/pull/4502/files#diff-a4b15b8f89712d3eeda4d7a79e9bc8720fff5807c606a63e1782f196f0eb96a8R59

File: src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1 [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    [Console]::InputEncoding = New-Object Text.UTF8Encoding $false 
+} 
+
+if (!$PSHOME) { 
+    $null = Get-Item -LiteralPath 'variable:PSHOME' 
+}
+else { 
+    Import-Module -Name ([System.IO.Path]::Combine($PSHOME, 'Modules\Microsoft.PowerShell.Management\Microsoft.PowerShell.Management.psd1')) 
+    Import-Module -Name ([System.IO.Path]::Combine($PSHOME, 'Modules\Microsoft.PowerShell.Utility\Microsoft.PowerShell.Utility.psd1')) 
+}
```
Comment (KonstantinTyukalov) [inline src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1:25]: [NotBlocking]: What does this mean actually? If $PSHome is not defined then try to get it and skip import modules?
Can't find any sense here but I see you copied that from the original script string so I would just to add `# TODO` for us to check that later.
Or if you have thoughts why this was added, please share with me🙂
Feel free to resolve.

File: src/dev.sh [hunk 5]
```diff
@@ -361,7 +361,16 @@ bash ./Misc/externals.sh $RUNTIME_ID "Pre-Cache" || checkRC "externals.sh Pre-Ca
...
+
+    # Tim Brigham - update to the default string first to search 
+    vs_location=$("$vswhere"  -latest -property installationPath)
+
+    # Then only if it is empty perform a second search wiith the BuildTools product set 
+    if [[ ! -e "${vs_location}" ]]; then
+        #try again with -products 'Microsoft.VisualStudio.Product.BuildTools' set 
+        vs_location=$("$vswhere"  -latest -property installationPath -products 'Microsoft.VisualStudio.Product.BuildTools')
+    fi
+
```
Comment (KonstantinTyukalov) [inline src/dev.sh:373]: Are these changes related to CLM support? 
If not, I would move this to a separate PR.

Comment (timbrigham-oc) [reply src/dev.sh]: I've disabled this change, will enter a separate PR to address in the future.

File: src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1 (hunk 4 above)
Comment (timbrigham-oc) [reply src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1]: Unknown as to the case, but not my logic. I've never seen pshome being left undefined. I copied the original logic over. I've added the todo mention to check at a later point.

File: src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1 (hunk 3 above)
Comment (timbrigham-oc) [reply src/Misc/layoutbin/powershell/Use-VstsTaskScript.ps1]: @KonstantinTyukalov, I've synced the names up. I have a habit of just using enough characters to disambiguate the name in my scripts, not the best to embed into a product. :)

File: src/Agent.Worker/Handlers/PowerShell3Handler.cs (hunk 2 above)
Comment (timbrigham-oc) [reply src/Agent.Worker/Handlers/PowerShell3Handler.cs]: Naming convention is updated, and I have updated the script path to use the well known directory.

**Review by timbrigham-oc — COMMENTED** [review]
I wasn't originally going to add instructions for use, but I had a chance to add detection statement for the current running language for both the Start-AzpTask script along with the default conditions on the machine. It should help identify the configuration issues that need to be resolved on hardened endpoints to use this in a production capacity.

**Comment by timbrigham-oc** [comment]
@KonstantinTyukalov, @DmitriiBobreshev, it's been a couple months.. 

What do I need to do to progress this request? I am assuming that the required check for azure-pipelines-agent.ci needs to be addressed. I'm seeing a single failure indicating that something timed out but nothing about the underlying cause or how to address it.

**Comment by timbrigham-oc** [comment]
@KonstantinTyukalov, @DmitriiBobreshev, why isn't this yet processed? Am I missing something?

**Comment by DenisNikulin5** [comment]
I'm closing this PR as these changes were merged here https://github.com/microsoft/azure-pipelines-agent/pull/4923
